presto-docs/src/main/sphinx/spill/overview.rst

```diff
@@ -0,0 +1,85 @@
+
+
+Memory management and spill
+---------------------------
+
+By default, Presto kills queries if the memory requested by the query execution
+exceeds session properties ``query_max_memory`` or
+``query_max_memory_per_node``. This mechanism ensures fairness in allocation
+memory to queries and prevents deadlock caused by memory allocation.
+It is efficient when there is a lot of small queries in the cluster, but
+leads to killing large queries that don't stay within the limits.
+
+To overcome this inefficiency, the concept of revocable memory was introduced. A
+query can request memory that does not count toward the limits, but this memory
+can be revoked by the memory manager at any time. When memory is revoked, the
+query runner spills intermediate data from memory to disk and continues to
+process it later.
+
+In practice, when the cluster is idle, a memory intensive query may use all of
+the memory in the cluster. On the other hand, when the cluster does not have
+much free memory, the same query may be forced to use disk as storage for
+intermediate data. A query that is forced to spill to disk may have a longer
+execution time by orders of magnitude than a query that runs completely in
+memory.
+
+Please note that enabling spill-to-disk does not guarantee execution of all
+memory intensive queries. It is still possible that the query runner will fail
+to divide intermediate data into chunks small enough that every chunk fits into
+memory, leading to `Out of memory` errors while loading the data from disk.
+
+Supported operations
+------------------------
+
+Not all operations support spilling to disk, and each handles spilling
+differently. Currently, the mechanism is implemented for the following
+operations.
+
+Joins
+^^^^^
+
+During join operation, memory is used to perform lookups for processed rows of
+joined tables. One of the joined tables is stored in memory and is used as
+lookup source for rows coming from the other table. The table stored in memory
+is called build table.
+
+If local concurrency is used for processing, the build table is partitioned into
+a number of partitions. The number of partitions is equal to the value of
+``task.concurrency`` configuration parameter (see :ref:`task-properties`).
+
+Having build table partitioned allows using spill-to-disk mechanism to decrease
+peak memory usage needed by join operator. When query approaches memory limit, a
+subset of partitions of the build table gets spilled to disk. From now on, rows
+from the other table that match to spilled build partitions are also spilled.
+This influences the amount of disk space needed.
+
+Afterward, the spilled partitions are read one-by-one to finish join operation.
+
+With this mechanism, the peak memory used by the join operator can be decreased
+to the size of the largest build table partition. Assuming no data skew, this will
+be 1/task.concurrency the size of the whole build table.
```

Comment:
"this will be 1/``task.concurrency`` times the size of the whole build table." or "this will be the size of the whole build table divided by ``task.concurrency``."